src/rt/profilegc.d

```diff
@@ -0,0 +1,151 @@
+ * switch
+ *
+ * Copyright: Copyright Digital Mars 2015 - 2015.
+ * License: Distributed under the
+ *      $(LINK2 http://www.boost.org/LICENSE_1_0.txt, Boost Software License 1.0).
+ *    (See accompanying file LICENSE)
+ * Authors:   Andrei Alexandrescu and Walter Bright
+ * Source: $(DRUNTIMESRC src/rt/_profilegc.d)
+ */
+
+module rt.profilegc;
+
+private:
+
+import core.stdc.stdio;
+import core.stdc.stdlib;
+import core.stdc.string;
+
+import core.exception : onOutOfMemoryError;
+
+char[] buffer;
+size_t[string] newCounts;
+
+__gshared
+{
+    size_t[string] globalNewCounts;
+    string logfilename = "profilegc.log";
+}
+
+/****
+ * Set file name for output.
+ * A file name of "" means write results to stdout.
+ * Params:
+ *      name = file name
+ */
+
+extern (C) void profilegc_setlogfilename(string name)
+{
+    logfilename = name;
+}
+
+
+
+public void accumulate(string file, uint line, string funcname, string type, size_t sz)
+{
+    char[3 * line.sizeof + 1] buf;
+    auto buflen = snprintf(buf.ptr, buf.length, "%u", line);
+
+    auto length = type.length + 1 + funcname.length + 1 + file.length + 1 + buflen;
+    if (length > buffer.length)
+    {
+        // Enlarge buffer[] so it is big enough
+        auto p = cast(char*)realloc(buffer.ptr, length);
+        if (!p)
+            onOutOfMemoryError();
+        buffer = p[0 .. length];
+    }
+
+    // "type funcname file:line"
+    memcpy(buffer.ptr, type.ptr, type.length);
+    buffer[type.length] = ' ';
+    memcpy(buffer.ptr + type.length + 1, funcname.ptr, funcname.length);
+    buffer[type.length + 1 + funcname.length] = ' ';
+    memcpy(buffer.ptr + type.length + 1 + funcname.length + 1, file.ptr, file.length);
+    buffer[type.length + 1 + funcname.length + 1 + file.length] = ':';
+    memcpy(buffer.ptr + type.length + 1 + funcname.length + 1 + file.length + 1, buf.ptr, buflen);
+
+    auto pcount = cast(string)buffer[0 .. length] in newCounts;
+    if (pcount)
+        *pcount += sz;                          // existing entry
+    else
+        newCounts[buffer[0..length].idup] = sz; // new entry
+
+    //string name = cast(string)(type ~ " " ~ funcname ~ " " ~ file ~ ":" ~ buf[0 .. buflen]);
+    //newCounts[name] += sz;
+}
+
+// Merge thread local newCounts into globalNewCounts
+static ~this()
+{
+    if (newCounts.length)
+    {
+        synchronized
+        {
+            if (globalNewCounts.length)
+            {
+                // Merge
+                foreach (name, count; newCounts)
+                {
+                    globalNewCounts[name] += count;
+                }
+            }
+            else
+                // Assign
+                globalNewCounts = newCounts;
+        }
+        newCounts = null;
+    }
+    free(buffer.ptr);
+    buffer = null;
+}
+
+// Write report to stderr
+shared static ~this()
+{
+    static struct Result
+    {
+        string name;
+        size_t count;
+
+        // qsort() comparator to sort by count field
```

Comment:
The hooks definitely belong in druntime, but it could do something like define an `interface` which phobos (or a third party) could implement.